operator/src/main/java/org/bf2/operator/managers/MetricsManager.java

```diff
@@ -43,68 +54,94 @@
-        meterRegistry.gauge(KAFKA_INSTANCE_PARTITION_LIMIT, tags, kafka, k -> kafkaConfigValue(k, KafkaCluster.MAX_PARTITIONS));
-        meterRegistry.gauge(KAFKA_INSTANCE_MAX_MESSAGE_SIZE_LIMIT, tags, kafka, k -> kafkaConfigValue(k, KafkaCluster.MESSAGE_MAX_BYTES));
-
-        if (kafka.getSpec() != null && kafka.getSpec().getKafka() != null) {
-            IntStream.range(0, kafka.getSpec().getKafka().getReplicas()).forEach(ordinal -> {
-                Tags brokerTags = Tags.concat(tags, Tags.of(Tag.of(TAG_LABEL_BROKER_ID, String.valueOf(ordinal))));
-
-                if (kafka.getSpec().getKafka().getListeners() != null) {
-                    kafka.getSpec().getKafka().getListeners().stream().filter(l -> EXTERNAL_LISTENER_NAME.equals(l.getName())).forEach(l -> {
-                        String listenerName = String.format("%s-%d", l.getName().toUpperCase(Locale.ROOT), l.getPort()); // Note - kafka itself capitalises the listener name.
-                        Tags listenerTags = Tags.concat(brokerTags, Tags.of(Tag.of(TAG_LABEL_LISTENER, listenerName)));
-                        meterRegistry.gauge(KAFKA_INSTANCE_CONNECTION_LIMIT, listenerTags, l, this::maxConnections);
-                        meterRegistry.gauge(KAFKA_INSTANCE_CONNECTION_CREATION_RATE_LIMIT, listenerTags, l, this::maxConnectionCreationRate);
-                    });
-                }
-            });
-        }
+
+        meterRegistry.gauge(KAFKA_INSTANCE_SPEC_BROKERS_DESIRED_COUNT, tags, ref, this::replicas);
+        meterRegistry.gauge(KAFKA_INSTANCE_PARTITION_LIMIT, tags, ref, k -> kafkaConfigValue(k, KafkaCluster.MAX_PARTITIONS));
+        meterRegistry.gauge(KAFKA_INSTANCE_MAX_MESSAGE_SIZE_LIMIT, tags, ref, k -> kafkaConfigValue(k, KafkaCluster.MESSAGE_MAX_BYTES));
+
+        Set<Meter> orphanMeters = new HashSet<>(Search.in(meterRegistry).tags(tags).tagKeys(TAG_LABEL_BROKER_ID).meters());
+
+        Optional<GenericKafkaListener> externalListener = getExternalListener(ref);
+        externalListener.ifPresent(genericKafkaListener -> IntStream.range(0, kafka.getSpec().getKafka().getReplicas()).forEach(ordinal -> {
+            String listenerName = String.format("%s-%d", genericKafkaListener.getName().toUpperCase(Locale.ROOT), genericKafkaListener.getPort()); // Note - kafka itself capitalises the listener name.
+            Tag brokerTag = Tag.of(TAG_LABEL_BROKER_ID, String.valueOf(ordinal));
+            Tag listenerTag = Tag.of(TAG_LABEL_LISTENER, listenerName);
+            Tags listenerTags = tags.and(brokerTag, listenerTag);
+
+            orphanMeters.removeIf(m -> m.getId().getTags().contains(brokerTag));
+
+            meterRegistry.gauge(KAFKA_INSTANCE_CONNECTION_LIMIT, listenerTags, ref,
+                    r -> getExternalListener(r).map(GenericKafkaListener::getConfiguration).map(GenericKafkaListenerConfiguration::getMaxConnections).map(Integer::doubleValue).orElse(Double.NaN));
```

Comment:
I'm new to this API - could you also pass `genericKafkaListener` (instead of `ref`) and skip the first step of `getExternalListener(r)` ?